### PR TITLE
Adcm 1842 ui tests: base page and login page refactoring

### PR DESCRIPTION
**Pull request by annademina**

### CODE REVIEW

File: tests/ui_tests/test_login_page.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    login_page.login_user(**adcm_credentials)
+    with allure.step("Check if user has been authorized"):
+        intro_page = AdminIntroPage(app_fs.driver, app_fs.adcm.url)
+        login_page.wait_url_contains_path(intro_page.path)
+        assert intro_page.path in app_fs.driver.current_url, \
+            f"Page '{intro_page.path}' has not been opened"
+        intro_page.header.check_auth_page_elements()
+
+
+@pytest.mark.parametrize("name, password", [("admin", " "), (" ", "admin")],
```
Comment (dgusakov) [inline tests/ui_tests/test_login_page.py:32]: No nothing?) It looks like we should be unable to press login if one of the fields is empty. Ask the web team about it.
Also need to add invalid user and invalid password cases

File: tests/ui_tests/test_login_page.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                         ids=("no_password", "no_name"))
+def test_check_error_in_login(app_fs, name, password):
+    params = {"error_text": "Incorrect password or user."}
+    login_page = LoginPage(app_fs.driver, app_fs.adcm.url).open()
+    login_page.login_user(name, password)
+    login_page.check_error_message(params["error_text"])
+
+
+def test_check_header_in_login_page(app_fs):
```
Comment (dgusakov) [inline tests/ui_tests/test_login_page.py:41]:
```suggestion
def test_check_header_links_in_login_page_unauthorised(app_fs):
```

File: tests/ui_tests/test_login_page.py (hunk 1 above)
Comment (annademina) [reply tests/ui_tests/test_login_page.py]: added two more tests with empty name and pass values and as a result - the unavailable login button

Comment (dgusakov) [reply tests/ui_tests/test_login_page.py]: Ok. My bad. I didn't notice a space between quotes. Consider changing ids and it will be ok!

**Review by dgusakov — APPROVED** [review]

**Review by vgorkavenko — APPROVED** [review]

File: tests/ui_tests/app/page/common/base_page.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+)
+
+
+class BasePageObject:
+    """BasePageObject is parent class for all ADCM's pages."""
+
+    def __init__(self, driver, base_url, path=""):
+        self.driver = driver
+        self.base_url = base_url
+        self.path = path
```
Comment (dgusakov) [inline tests/ui_tests/app/page/common/base_page.py:38]:
```suggestion
    __slots__ = {"driver", "base_url", "path", "default_wait_timeout"}

    def __init__(self, driver, base_url, path="", default_wait_timeout=5):
        self.driver = driver
        self.base_url = base_url
        self.path = path
        self.default_wait_timeout = default_wait_timeout
```

File: tests/ui_tests/app/page/common/base_page.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    def open(self, timeout=10):
+        """Open page by its url and path."""
+
+        url = self.base_url + self.path
+        if self.driver.current_url != url:
+            with allure.step(f"Open {url}"):
+                self.driver.get(url)
+        self.wait_url_contains_path(self.path, timer=timeout)
+        return self
```
Comment (dgusakov) [inline tests/ui_tests/app/page/common/base_page.py:48]:
```suggestion
    def open(self, timeout=None):
        """Open page by its url and path."""

        url = self.base_url + self.path
        if self.driver.current_url != url:
            with allure.step(f"Open {url}"):
                self.driver.get(url)
        self.wait_url_contains_path(self.path, timer=timeout or self.default_wait_timeout)
        return self
```

Comment (dgusakov) [reply tests/ui_tests/app/page/common/base_page.py:48]: Do the same for the other methods with timeouts

File: tests/ui_tests/app/page/login/login_page.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            LoginPageLocators.login_btn,
+        ])
+        self.footer.check_all_elements()
+
+    @allure.step('Get warning text on login page')
+    def get_login_warning_text(self) -> str:
+        def get_text():
+            assert self.find_element(LoginPageLocators.login_warning).text != ""
+        wait_until_step_succeeds(get_text)
+        return self.find_element(LoginPageLocators.login_warning).text
```
Comment (dgusakov) [inline tests/ui_tests/app/page/login/login_page.py:47]:
```suggestion
    def get_login_warning_text(self, timeout=None) -> str:
        def get_text():
            assert self.find_element(LoginPageLocators.login_warning).text != ""
        wait_until_step_succeeds(get_text, period=1, timeout=timeout or self.default_wait_timeout)
        return self.find_element(LoginPageLocators.login_warning).text
```

**Review by dgusakov — CHANGES_REQUESTED** [review]

File: tests/ui_tests/app/helpers/locator.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+from selenium.webdriver.common.by import By
+
+
+@dataclass
+class Locator:
+    """Describes a locator on a webpage"""
+
+    by: By
+    path: str = ""
+    name: str = ""
```
Comment (vgorkavenko) [inline tests/ui_tests/app/helpers/locator.py:23]: I think these fields always need to object creation. Maybe remove the default field value?

File: tests/ui_tests/app/helpers/locator.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+from dataclasses import dataclass
+from selenium.webdriver.common.by import By
+
+
+@dataclass
+class Locator:
+    """Describes a locator on a webpage"""
+
+    by: By
+    path: str = ""
```
Comment (vgorkavenko) [inline tests/ui_tests/app/helpers/locator.py:22]: It seems to me, `path` is related to `xpath` search strategy. But we have not only `xpath`. I think need more universality for the field name

File: tests/ui_tests/app/page/common/base_page.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        self.path = path
+        self.default_wait_timeout = default_wait_timeout
+
+    def open(self, timeout=None):
+        """Open page by its url and path."""
+
+        url = self.base_url + self.path
+        if self.driver.current_url != url:
+            with allure.step(f"Open {url}"):
+                self.driver.get(url)
```
Comment (vgorkavenko) [inline tests/ui_tests/app/page/common/base_page.py:49]: Should we assign a new value for `self.driver.current_url` after `self.driver.get(url)`? It seems to me if we don't, that always re-open the same page

File: tests/ui_tests/app/page/common/base_page.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+    def find_elements(self, locator: Locator, time: int = 15) -> [WebElement]:
+        """Find elements on current page."""
+
+        with allure.step(f'Find elements "{locator.name}" on page'):
+            return WDW(self.driver, time).until(EC.presence_of_all_elements_located([locator.by, locator.path]),
+                                                message=f"Can't find {locator.name} on page "
+                                                        f"{self.driver.current_url} for {time} seconds")
+
+    def is_locator_displayed(self, locator: Locator, time: int = 10) -> bool:
+        """Checks if locator is displayed."""
```
Comment (vgorkavenko) [inline tests/ui_tests/app/page/common/base_page.py:76]: Hereinafter. We search element, not locator. I think is better if the names of functions will contain `element`, not `locator`. 
It might be worth considering naming `Locator` class as well. It seems that the meaning and content say us that is `Element`

File: tests/ui_tests/app/helpers/locator.py (hunk 7 above)
Comment (annademina) [reply tests/ui_tests/app/helpers/locator.py:22]: like value?

File: tests/ui_tests/app/helpers/locator.py (hunk 6 above)
Comment (annademina) [reply tests/ui_tests/app/helpers/locator.py]: good idea, removed from name and path

File: tests/ui_tests/app/helpers/locator.py (hunk 7 above)
Comment (vgorkavenko) [reply tests/ui_tests/app/helpers/locator.py]: Actually, I don't have any good idea :) Mb `locator_value` ? 
@dgusakov what do you think?

File: tests/ui_tests/app/page/common/base_page.py (hunk 8 above)
Comment (annademina) [reply tests/ui_tests/app/page/common/base_page.py:67]: no, it's a property of selenium WebDriver and each time it will execute the command get_current_url

File: tests/ui_tests/app/page/common/base_page.py (hunk 9 above)
Comment (annademina) [reply tests/ui_tests/app/page/common/base_page.py]: renamed methods

File: tests/ui_tests/app/helpers/locator.py (hunk 7 above)
Comment (annademina) [reply tests/ui_tests/app/helpers/locator.py]: navigation/locating value? definition? identificator?

**Review by dgusakov — DISMISSED** [review]

File: tests/ui_tests/app/helpers/locator.py (hunk 7 above)
Comment (annademina) [reply tests/ui_tests/app/helpers/locator.py]: changed to value

**Review by dgusakov — APPROVED** [review]

**Review by vgorkavenko — DISMISSED** [review]

**Review by dgusakov — APPROVED** [review]

**Review by vgorkavenko — APPROVED** [review]
